perf/src/qos.rs

```diff
@@ -0,0 +1,121 @@
+use {
+    crate::{
+        data_budget::DataBudget,
+        packet::{Packet, PacketBatch},
+    },
+    std::{
+        collections::HashMap,
+        net::IpAddr,
+        sync::atomic::{AtomicUsize, Ordering},
+    },
+};
+
+pub type StakeMap = HashMap<IpAddr, (u64, DataBudget)>;
+
+fn qos_packet(
+    num_discarded: &AtomicUsize,
+    max_packets_per_second: u64,
+    default_budget: &DataBudget,
+    total_stake: u64,
+    ip_to_budget: &StakeMap,
+    packet: &mut Packet,
+) {
+    if packet.meta.discard() {
+        return;
+    }
+
+    if let Some((stake, budget)) = ip_to_budget.get(&packet.meta.addr) {
+        budget.update(1000, |_| {
+            (*stake as u128 * max_packets_per_second as u128 / total_stake as u128) as usize
```

Comment:
why not just use a float?  too slow?

Reply:
I tried just u64 and the speed was the same. Yea float32 is probably the best in theory, but I'm not sure a few integer instructions is changing the speed much.

Reply:
tried float32, slower than u128 on intel cascade lake